crates/ra_hir/src/source_analyzer.rs

```diff
@@ -305,6 +310,49 @@ impl SourceAnalyzer {
         self.resolve_hir_path(db, &hir_path)
     }
 
+    fn resolve_as_full_path(&self, path_expr: ast::PathExpr) -> Option<PathResolution> {
+        let expr_id = self.expr_id(&path_expr.into())?;
+        self.infer
+            .as_ref()?
+            .assoc_resolutions_for_expr(expr_id)
+            .map(|assoc| PathResolution::AssocItem(assoc.into()))
+    }
+
+    fn resolve_as_path_to_method(
+        &self,
+        db: &impl HirDatabase,
+        path_expr: &ast::PathExpr,
+    ) -> Option<PathResolution> {
+        let full_path = path_expr.path()?;
+        let path_to_method = full_path.qualifier()?;
+        let method_name = full_path.segment()?.syntax().to_string();
+        match self.resolve_path(db, &path_to_method)? {
+            PathResolution::Def(ModuleDef::Adt(adt)) => {
+                let ty = adt.ty(db);
+                iterate_method_candidates(
```

Comment:
Why do we need to do this here? 

I would expect that we reoslve assoc functions during type inferece, and record the result in the appropriate map in `InferenceResult`. Perhaps `assoc_resolutions` are not properly filled in?

Reply:
That actually comes from a very confusing (to me) SourceAnalyzer behavior: when constructed for a module (as it had been done before in auto_import assist code: https://github.com/rust-analyzer/rust-analyzer/pull/3120/files#diff-9f55419f762ae3317cf277aab2d8e37aL36), it does not resolve anything related to the functions' types.

If constructed for a very small path under the coursor, it magically starts to resolve those types.

I've removed the module syntax search completely since it was redundant, but did not realize that this code is not needed anymore.

Reply:
This behavior depends on what the `source_binder::find_container` function returns and I find it a bit strange, I would expect to be able to resolve all types, including the function-related ones when the whole module code is give to the analyzer.

Reply:
Yeah, this aspect of SourceAnalyzer is kind of a mess, you need to use the right source analyzer. We really should do something like lifting all sa methods to sb, such that they just work.

Not sure how to make that ergonomic though, because the caller would have to pass in `InFile<ast::Foo>` things instead of just `ast::Foo`.

Reply:
Yeah, this is something that has annoyed me as well. It would be nice if working with `InFile<>`s would be more ergonomic in general, and then we could maybe put more API directly on them :thinking:

Reply:
Just curious: what prevents the analyzer from traversing inner elements and extracting additional info?
That what confuses me: when I pass more info to the analyzer, it gives me less data in return yet when I pass a smaller subset of information, it starts to work.
So on both cases there's the same ast part available, that is enough to produce the data I need.

Reply:
That's not less information: syntax trees have parent links, so any node knows about the whole file. It wouldn't be too hard to make sure that it "just works" for all syntax elements from the single file. The problems is that, ideally, you don't want to care about which file are you in (b/c with macros you generally hop between files even within a single function), but SyntaxNodes kind of purpusfully don't know the file.

Reply:
Thanks for the explanation.

Reply:
bwaaaa, why dealing with macros in IDE is so annoying :-( 

If I just go with the most obvious API, I get this for, for example, goto type definition:

![image](https://user-images.githubusercontent.com/1711539/74360876-24885800-4dc6-11ea-9ead-f8903996aad6.png)

That's way to many layers of Functors  :c

Reply:
Isn't `and_then` a monad?

Reply:
We could at least e.g. add a `cast<A: AstNode>(&self: InFile<SyntaxNode>) -> Option<InFile<A>>`...

Reply:
Oh, right, we do need full monadic power here :-( 

And yeah, for full generality we need ReaderT monad transformer ;(

I wish we could do a simple Roslyn style api: https://github.com/dotnet/roslyn/wiki/Getting-Started-C%23-Semantic-Analysis